src/Algebra/Bundles/Free/Magma.agda

```diff
@@ -0,0 +1,443 @@
+
+  open Magma 𝓜
+    renaming (Carrier to UM; _≈_ to _≈ᴹ_; _∙_ to _∙ᴹ_
+             ; setoid to setoidᴹ; rawMagma to rawMagmaᴹ
+             ; isMagma to isMagmaᴹ)
+
+  open ≈-Reasoning setoidᴹ
+
+  open Syntax
+
+  open Setoid 𝓐 renaming (Carrier to UA; _≈_ to _≈ᴬ_)
+
+  open Eval 𝓜 𝓐 public
+
+  open FreeMagma 𝓐 renaming (setoid to FA; Carrier to UFA)
+
+  open SetoidHomomorphism 𝓗 renaming (⟦_⟧ to η; isRelHomomorphism to hom-η) 
+
+  private
+  
+    ⟦_⟧ᴹ : UFA → UM
+    ⟦_⟧ᴹ = ⟦_⟧ η
+
+  open Structures _≈ᴹ_
+  open IsMagma isMagmaᴹ renaming (∙-cong to congᴹ)
+  open IsRelHomomorphism hom-η renaming (cong to cong-η)
+    
+  module Existence where
+  
+    private
+      algᴹ = alg 𝓜
+
+    unfold-⟦_⟧ᴹ : ∀ t → ⟦ t ⟧ᴹ ≈ᴹ algᴹ (map η t)
+    unfold-⟦ var a ⟧ᴹ = begin η a ∎
+    unfold-⟦ s ∙ t ⟧ᴹ = congᴹ unfold-⟦ s ⟧ᴹ unfold-⟦ t ⟧ᴹ
+
+    cong-⟦_⟧ : ∀ {s t} → s ≈ t → ⟦ s ⟧ᴹ ≈ᴹ ⟦ t ⟧ᴹ
+    cong-⟦ var r ⟧ = cong-η r
+    cong-⟦ s ∙ t ⟧ = congᴹ cong-⟦ s ⟧ cong-⟦ t ⟧
+
+    isRelHomomorphismᴹ : IsRelHomomorphism _≈_ _≈ᴹ_ ⟦_⟧ᴹ
+    isRelHomomorphismᴹ = record { cong = cong-⟦_⟧ }
+
+    setoidHomomorphismᴹ : SetoidHomomorphism FA setoidᴹ
+    setoidHomomorphismᴹ = record { ⟦_⟧ = ⟦_⟧ᴹ ; isRelHomomorphism = isRelHomomorphismᴹ }
+
+    isMagmaHomomorphismᴹ : IsMagmaHomomorphism rawMagma rawMagmaᴹ ⟦_⟧ᴹ
+    isMagmaHomomorphismᴹ = record { isRelHomomorphism = isRelHomomorphismᴹ
+                                  ; homo = λ s t → begin ⟦ s ⟧ᴹ ∙ᴹ ⟦ t ⟧ᴹ ∎ }
+
+    magmaHomomorphismᴹ : MagmaHomomorphism freeMagma 𝓜
+    magmaHomomorphismᴹ = record { ⟦_⟧ = ⟦_⟧ᴹ
+                               ; isMagmaHomomorphism = isMagmaHomomorphismᴹ }
+
+  record η-MagmaHomomorphism : Set (suc (a ⊔ m ⊔ ℓa ⊔ ℓm)) where
+
+    field
+      magmaHomomorphism : MagmaHomomorphism freeMagma 𝓜
+    open MagmaHomomorphism magmaHomomorphism public renaming (homo to ⟦⟧-homo)
+    field
+      ⟦_⟧∘var≈ᴹη : ∀ a → ⟦ var a ⟧ ≈ᴹ η a
+
+  ⟦⟧ᴹ-η-MagmaHomomorphism : η-MagmaHomomorphism
+  ⟦⟧ᴹ-η-MagmaHomomorphism = record { magmaHomomorphism = Existence.magmaHomomorphismᴹ
+                                   ; ⟦_⟧∘var≈ᴹη = Existence.unfold-⟦_⟧ᴹ ∘ var } 
+  
+  module Uniqueness (η-magmaHomomorphism : η-MagmaHomomorphism) where
+      
+    open η-MagmaHomomorphism η-magmaHomomorphism
+      
+    isUnique⟦_⟧ᴹ : ∀ t → ⟦ t ⟧ ≈ᴹ ⟦ t ⟧ᴹ
+    isUnique⟦ var a ⟧ᴹ = ⟦ a ⟧∘var≈ᴹη
+    isUnique⟦ s ∙ t ⟧ᴹ = begin
+        ⟦ s Syntax.∙ t ⟧  ≈⟨ ⟦⟧-homo s t ⟩
+        ⟦ s ⟧ ∙ᴹ ⟦ t ⟧    ≈⟨ congᴹ isUnique⟦ s ⟧ᴹ isUnique⟦ t ⟧ᴹ ⟩
+        ⟦ s ⟧ᴹ ∙ᴹ ⟦ t ⟧ᴹ  ∎
+
+  module Corollary (𝓗 𝓚 : η-MagmaHomomorphism)
+    where
+      open η-MagmaHomomorphism 𝓗 renaming (⟦_⟧ to ⟦_⟧ᴴ)
+      open η-MagmaHomomorphism 𝓚 renaming (⟦_⟧ to ⟦_⟧ᴷ)
+      open Uniqueness 𝓗 renaming (isUnique⟦_⟧ᴹ to isUnique⟦_⟧ᴴ)
+      open Uniqueness 𝓚 renaming (isUnique⟦_⟧ᴹ to isUnique⟦_⟧ᴷ)
+      
+      isUnique⟦_⟧ :  ∀ t → ⟦ t ⟧ᴴ ≈ᴹ ⟦ t ⟧ᴷ
+      isUnique⟦ t ⟧ = begin ⟦ t ⟧ᴴ ≈⟨ isUnique⟦ t ⟧ᴴ ⟩ ⟦ t ⟧ᴹ ≈˘⟨ isUnique⟦ t ⟧ᴷ ⟩ ⟦ t ⟧ᴷ ∎
```

Comment:
maybe do a proper vertical layout for this equational reasoning?

Reply:
Again, here, because it really is an equation of the form `trans eq1 (sym eq2)` I though it easier to inline it. But maybe producing a custom combinator for this case would be clearer?. It doesn't feel elaborate enough to rquire a full Feijen-style layout, though.

Reply:
But without that layout it's very difficult to tell it's actually a `trans ... (sym ...)`!

Reply:
If you're concerned about vertical space I'd consider writing it explicitly using `trans`

Reply:
I've decided to leave it as it is.